Number_types/include/CGAL/cpp_float.h

```diff
@@ -0,0 +1,496 @@
+  }
+
+
+
+  cpp_float operator-=(const cpp_float& other)
+  {
+
+#ifdef CGAL_CPPF
+    rat -= other.rat;
+#endif
+    int shift = exp - other.exp;
+    if(shift > 0){
+      man <<= shift;
+      man -= other.man;
+      exp = other.exp;
+    }else if(shift < 0){
+      man -= (other.man << -shift);
+    }else{
+      man -= other.man;
+    }
+    return *this;
+  }
+
+  friend
+  cpp_float operator-(const cpp_float& a, const cpp_float&b){
+
+    int shift = a.exp - b.exp;
+    if(shift > 0){
+      return cpp_float((a.man << shift) - b.man, b.exp);
+    }else if(shift < 0){
+      return cpp_float(a.man - (b.man << -shift), a.exp);
+    }
+    return cpp_float(a.man - b.man, a.exp);
+  }
+
+  bool positive() const
+  {
+    return is_positive(man);
+  }
+
+
+  friend bool operator<(const cpp_float& a, const cpp_float& b)
+  {
+#ifdef CGAL_CPPF
+    bool qres = a.rat < b.rat;
+#endif
+    cpp_float d(b);
+    d -= a;
+
+#ifdef CGAL_CPPF
+    assert(qres == ( (!d.is_zero()) && d.positive()));
+#endif
+    return ( (!d.is_zero()) && d.positive());
+  }
+
+  friend bool operator>(cpp_float const&a, cpp_float const&b){
+    return b<a;
+  }
+  friend bool operator>=(cpp_float const&a, cpp_float const&b){
+    return !(a<b);
+  }
+  friend bool operator<=(cpp_float const&a, cpp_float const&b){
+    return !(a>b);
+  }
+
+
+  friend bool operator==(cpp_float const&a, cpp_float const&b){
+
+#ifdef CGAL_CPPF
+    bool qres = a.rat == b.rat;
+#endif
+   int shift = a.exp - b.exp;
+    if(shift > 0){
+      Mantissa ac(a.man);
+      ac <<= shift;
+#ifdef CGAL_CPPF
+      assert( qres == (ac == b.man));
+#endif
+      return ac == b.man;
+    }else if(shift < 0){
+      Mantissa  bc(b.man);
+      bc <<= -shift;
+#ifdef CGAL_CPPF
+      assert(qres == (a.man == bc));
+#endif
+      return a.man == bc;
+    }
+#ifdef CGAL_CPPF
+    assert(qres == (a.man == b.man));
+#endif
+    return a.man==b.man;
+  }
+
+  Comparison_result compare(const cpp_float& other) const
+  {
+    if(*this < other) return SMALLER;
+    if(*this > other) return LARGER;
+    return EQUAL;
+  }
+
+  friend bool operator!=(cpp_float const&a, cpp_float const&b){
+    return !(a==b);
+  }
+
+  double to_double() const
+  {
+    if(exp == 0){
+      return CGAL::to_double(man);
+    }
+    if(exp > 0){
+      Mantissa as(man);
+      as <<= exp;
+      return CGAL::to_double(as);
+    }
+    Mantissa pow(1);
+    pow <<= -exp;
+    boost::multiprecision::cpp_rational rat(man, pow);
+    return CGAL::to_double(rat);
+  }
+
+  std::pair<double,double> to_interval() const
+  {
+    assert(false);
+    double zero(0);
+    return std::make_pair(zero,zero);
+  }
+
+
+  bool is_zero () const {
+    return man==0 && exp == 0;
+  }
+
+
+  bool is_one () const {
+    assert(false);
+    return true;
+    //  return exp==0 && size==1 && data()[0]==1;
+  }
+
+
+  CGAL::Sign sign () const
+  {
+    return CGAL::sign(man);
+  }
+
+};
+
+
+  template <> struct Algebraic_structure_traits< cpp_float >
```

Comment:
@mglisse I copied this from `Mpzf`, but I am wondering if we should only keep what is required by `Integral_domain_without_division_tag`.